src/control/server/mgmt_system.go

```diff
@@ -1072,6 +1073,148 @@ func (svc *mgmtSvc) SystemExclude(ctx context.Context, req *mgmtpb.SystemExclude
 	return resp, nil
 }
 
+func (svc *mgmtSvc) SystemDrain(ctx context.Context, req *mgmtpb.SystemDrainReq) (*mgmtpb.SystemDrainResp, error) {
+	if err := svc.checkLeaderRequest(wrapCheckerReq(req)); err != nil {
+		return nil, err
+	}
+
+	if req.Hosts == "" && req.Ranks == "" {
+		return nil, errors.New("no hosts or ranks specified")
+	}
+
+	hitRanks, missRanks, missHosts, err := svc.resolveRanks(req.Hosts, req.Ranks)
+	if err != nil {
+		return nil, err
+	}
+
+	if missHosts.Count() > 0 {
+		return nil, errors.Errorf("invalid host(s): %s", missHosts.String())
+	}
+	if missRanks.Count() > 0 {
+		return nil, errors.Errorf("invalid rank(s): %s", missRanks.String())
+	}
+	if hitRanks.Count() == 0 {
+		return nil, errors.New("no ranks to drain")
+	}
+
+	// Drain rank on each pool it belongs to.
+
+	psList, err := svc.sysdb.PoolServiceList(false)
+	if err != nil {
+		return nil, err
+	}
+
+	ranksMap := make(map[ranklist.Rank]struct{})
+	for _, r := range hitRanks.Ranks() {
+		ranksMap[r] = struct{}{}
+	}
+
+	poolRanks := make(map[string]*ranklist.RankSet)
+	poolIDs := []string{} // Label or UUID.
+	for _, ps := range psList {
+		currentRanks := ps.Storage.CurrentRanks()
+
+		// Label preferred over UUID.
+		poolID := ps.PoolLabel
+		if poolID == "" {
+			poolID = ps.PoolUUID.String()
+		}
+
+		svc.log.Tracef("pool-service detected: id %s, ranks %v", poolID, currentRanks)
+
+		for _, r := range currentRanks {
+			if _, exists := ranksMap[r]; !exists {
+				continue
+			}
+			if _, exists := poolRanks[poolID]; !exists {
+				poolRanks[poolID] = ranklist.MustCreateRankSet("")
+				poolIDs = append(poolIDs, poolID)
+			}
+			poolRanks[poolID].Add(r)
+		}
+	}
+	svc.log.Debugf("pool-ranks to drain: %v", poolRanks)
+
+	sort.Strings(poolIDs)
+
+	resp := new(mgmtpb.SystemDrainResp)
+	drainReq := new(mgmtpb.PoolDrainReq)
+	drainReq.Sys = req.Sys
+
+	for _, id := range poolIDs {
+		rs := poolRanks[id]
+		if rs.Count() == 0 {
+			continue
+		}
+		drained := ranklist.MustCreateRankSet("")
+		failed := make(map[string]*ranklist.RankSet)
+
+		// Use our incoming request and just replace relevant parameters on each iteration.
+		drainReq.Id = id
+
+		// TODO DAOS-6611: Drain multiple pool-ranks per call when drpc.MethodPoolDrain API
+		//                 supports it.
+		for _, r := range rs.Ranks() {
+			var errMsg string
+			drainReq.Rank = r.Uint32()
+
+			drpcResp, err := svc.makeLockedPoolServiceCall(ctx, drpc.MethodPoolDrain,
+				drainReq)
+			if err != nil {
+				return nil, err
+			}
+
+			drainResp := &mgmtpb.PoolDrainResp{}
+			if err = proto.Unmarshal(drpcResp.Body, drainResp); err != nil {
+				errMsg = errors.Wrap(err, "unmarshal PoolEvict response").Error()
+				drainResp.Status = int32(daos.IOInvalid)
```

Comment:
In dRPC we have an error code for this kind of failure: `drpc.UnmarshalingPayloadFailure()`

IMO when aggregating these results, we really want to flag cases like this as dRPC communication errors, rather than making them look like daos_engine errors.

Reply:
agreed, I didn't add this code, just edited, will update if I have to repush